README.md
LIB-POSIX2
==========

POSIX, an acronym for "Portable Operating System Interface", is a family of standards specified by the IEEE for maintaining compatibility between operating systems.


LICENSE
===============
* GNU GPL V2
* BSD

COMPILE TOOLS
===============
* 

AUTHORS
===============
* Arnd Hanses,
* Alexander Mai,
* Holger Veit

LINKS
===============
* http://posix2.sourceforge.net
* http://pubs.opengroup.org/onlinepubs/9699919799/


ChangeLog
		[$Id: ChangeLog,v 1.34 2000/11/27 18:13:42 arnd Exp $]
 
2000/03/14 LIBREVISION 0.01
	Initial version (arnd)
2000/03/22
	added fixcrcr utility and missing Makefiles in libext/uni (veit)

2000/03/24 LIBREVISION 0.02
	big change: (arnd)
	beautified fixcrcr.c, more docs, added syslog/drand/valloc/dlfcn,
	more headers, added some functions from EMX stat.c, added some
	i387 libm asm funcs, added a math test vector
2000/03/29 
	additions: (arnd)
	signal,resource,time header files and man pages
	Updated CodeRules
	Added Id lines to various files
	Support for itimer and daemon
	Misc fixes

2000/04/03 LIBREVISION 0.03
	additions: (arnd)
	more BSD man2 docs, signal(3) docs
	some missing headers (limits, signal, cdefs, itimer, resource)
	signalx.c was missing in CVS
	popen.c
	fixed a few Makefiles
	more CVSIds in files

2000/04/17 LIBREVISION 0.04 (0.04a)
	additions: (arnd)
	more docs, several sys headers from BSD
	regex support, some more stdlib: a64l, quad, tsearch
	sysconf support
	test for popen
	renamed fixcrcr.exe to fixtxt.exe
2000/04/18 (0.04b)
	additions: (arnd)
	err* stuff, __volatile__ macro fixed
	several minor CVS-Id changes
	msginfo structure renamed (clash with macro)
	removed timext.h
	_getPosixOS2Root cast fixed

2000/04/26 LIBREVISION 0.05
	additions: (arnd)
	perror, vis, man pages
	declare several arrays as pointers rather than []
	various conformance tests for FP and regex
	finally remove timext.h :-)
	set several constants from param.h which where hardcoded before
	init some local variables in regcomp to NULL
	empty getloadavg
	added some assembler math functions (i387math.S) from libgcc
	regexp debug function

2000/05/02 LIBREVISION 0.06
	additions: (arnd, Stefan Neis, Ralf Barkow)
	i387math.S fixes for acos/asin/exp from FreeBSD (neis)
	move fcntl.h after sys/types.h (barkow) and sys/stat.h (arnd)
	man pages: poll,bitstring,cat*,get*,killpg,math,hz,intro9
	i386 specific float.h, ieee.h->ieeefp.h
	nls support,killpg declaration, 
	IPC and SHM fixed,zlib integration
	mmap from BSD removed, nls messages for different languages
	sysconf fixes
	tests for db,shm
2000/05/05
	Math update (Cephes Math Lib), addition of SYS-facitilities.txt
	moved drand
	don't use gcc builtins
	handling of infinity
	test for floatingpoint funcs, consistency test
2000/05/08
	Math update s_tanh.c (neis)
2000/05/12
	Fix to stdio and unistd headers to make C++ happy (neis)

2000/05/15 LIBREVISION 0.07
	additions: ctype, libgen,
	docs for mmap, man5 pages
	tools: the ultimate editor (porting example), pmake
2000/05/17
	Minor update (arnd)
	additions: 4.3 compatible regex, realpath, nice
	more manpages
	HowTo pages for some OS/2 stuff
2000/05/31
	Update rev007b (arnd)
	fixcrcr.c: added check to detect binaries (\0 bytes in text)
	changed librevision format in dll.c
	added several EDM/2 articles and moved html files to docs/html
	some more manpages
	More stuff to FAQ
	renamed some fdlibm functions to __fdlibm
	__STRICT_ANSI__ removed from some defines
	register in prototypes removed
	itimer management improved
	added getsubopt,l64a,mergesort,bm,strlcpy,strmode,strsep
	non-Posix string extensions: strcasecmp etc.
2000/06/13
	math.h _LIB_VERSION var moved (neis)
2000/07/06
	minor changes of defs in math.h and limits.h for LONG LONG support
	(arnd)

Alpha #1 released

2000/07/24
	missing stuff for tools/pmake (arnd)
2000/07/25
	missing newline character added in some headers (amai)
	OpenBSD version of db added (arnd)
	misc fixes: headers, sources, docs (arnd)
2000/07/28
	<poll.h> added, minor updates for docs (amai)

2000/07/29 LIBREVISION 0.09
	lockf() added: not functional yet (see fcntl() in EMX) (arnd)
	timezone stuff added (arnd)
        Minor overall fixed (Makefiles, docs) (amai)

Alpha #2 released

2000/08/06
	Update rev009b
	time.h added, mmap fixes (arnd)
	removed #warnings from some headers - will make autoconf happy (arnd)
2000/08/08
	Update rev009c
	signgam added to libm (arnd)
	plain text version of FAQ removed in favour of HTML version (arnd)
	Makefile fixes (arnd, amai)
	[minor fixes to Makefiles and documentation are not always
	 logged here!]

2000/08/14 LIBREVISION 0.0.10
	misc fixes for libext (move some stuff in according subdirectories,
	dirent stuff, ...)   (arnd)
	Add big tree of utilities to tools/
2000/08/14 LIBREVISION 0.0.11
	few minor fixes for libExt (arnd)
	Modifications for Makefiles, make use of .mk files (arnd)
	"Missing" stuff in tools/ from rev 010 (arnd)
	working copy of rename(); tests for rename() and stat() (amai)
2000/08/15 LIBREVISION 0.0.11a
	few minor fixes for libExt; cleanup of directories (arnd)
	add data files to src/share (arnd)
	misc fixes for tools/ tree; added find(1) (arnd)
2000/08/18 LIBREVISION 0.0.11b
	Misc changes and additions; more files moved location (arnd)
	(In an early state this release was messed up in CVS!)
2000/08/19 LIBREVISION 0.0.12
	Misc stuff as usual. Mainly Makfiles this time,
	introduce two dlls instead of the single one
      change directory structures a bit (arnd)
2000/08/19 LIBREVISION 0.0.12a
2000/08/21 LIBREVISION 0.0.12b
2000/08/23 LIBREVISION 0.0.12c
	Dirent headers, glob* + misc stuff (arnd)
2000/08/26 LIBREVISION 0.0.13
	Build up basic directory structures (arnd)
      Add calendar and rmdir to tools/ as well as some manpages (arnd)
      Fixes/additions to libext
2000/09/09 LIBREVISION 0.0.14
      Add asa and printenv to tools/ as well as some manpages and 
      docs (arnd).
      Fixes/additions to libExt (arnd):
	1) New directory libext/libc/sys for modules that emulate BSD system 
	calls in a safer and more compatible way. Moved several ones.
	2) sys/mm.c enhanced and prepared for later split that will be 
	necessary for linking into a run-time DLL. New function 
	mprotect(). Now mmap() works with MAP_SHARED.
	3) sys/statx.c enhanced: New function statfs(). Several headers 
	adapted: sys/vfs.h, sys/statfs.h, sys/mount.h.
2000/10/11 LIBREVISION 0.0.14a (arnd)
	more changes of mm.c (MAP_SHARED/MAP_FIXED better supported)

tools/cp/cp.c
/*	$Id: cp.c,v 1.2 2000/08/14 16:26:08 amai Exp $	*/
/*	$NetBSD: cp.c,v 1.14 1995/09/07 06:14:51 jtc Exp $	*/

/*
 * Copyright (c) 1988, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * David Hitz of Auspex Systems Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgment:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
static char copyright[] =
"@(#) Copyright (c) 1988, 1993, 1994\n\
	The Regents of the University of California.  All rights reserved.\n";
static char sccsid[] = "@(#)cp.c	8.5 (Berkeley) 4/29/95";
static char rcsid[] = "$OpenBSD: cp.c,v 1.13 1999/05/06 17:19:45 millert Exp $";
#endif

/*
 * Cp copies source files to target files.
 * 
 * The global PATH_T structure "to" always contains the path to the
 * current target file.  Since fts(3) does not change directories,
 * this path can be either absolute or dot-relative.
 * 
 * The basic algorithm is to initialize "to" and use fts(3) to traverse
 * the file hierarchy rooted in the argument list.  A trivial case is the
 * case of 'cp file1 file2'.  The more interesting case is the case of
 * 'cp file1 file2 ... fileN dir' where the hierarchy is traversed and the
 * path (relative to the root of the traversal) is appended to dir (stored
 * in "to") to form the final target path.
 */

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/time.h>

#include <dirent.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <fts.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "extern.h"

#define	STRIP_TRAILING_SLASH(p) {					\
	while ((p).p_end > (p).p_path + 1 && (p).p_end[-1] == '/')	\
		*--(p).p_end = '\0';					\
}

PATH_T to = { to.p_path, "" };

uid_t myuid;
int Rflag, fflag, iflag, pflag, rflag;
int myumask;

enum op { FILE_TO_FILE, FILE_TO_DIR, DIR_TO_DNE };

int copy __P((char *[], enum op, int));
int mastercmp __P((const FTSENT **, const FTSENT **));
char *find_last_component __P((char *));

int
main(int argc, char *argv[])
{
	struct stat to_stat, tmp_stat;
	enum op type;
	int Hflag, Lflag, Pflag, ch, fts_options, r;
	char *target;
	Hflag = Lflag = Pflag = Rflag = 0;

	__progname = argv[0];
	_wildcard(&argc, &argv);

	while ((ch = getopt(argc, argv, "HLPRfipr")) != -1)
		switch (ch) {
		case 'H':
			Hflag = 1;
			Lflag = Pflag = 0;
			break;
		case 'L':
			Lflag = 1;
			Hflag = Pflag = 0;
			break;
		case 'P':
			Pflag = 1;
			Hflag = Lflag = 0;
			break;
		case 'R':
			Rflag = 1;
			break;
		case 'f':
			fflag = 1;
			iflag = 0;
			break;
		case 'i':
			iflag = isatty(fileno(stdin));
			fflag = 0;
			break;
		case 'p':
			pflag = 1;
			break;
		case 'r':
			rflag = 1;
			break;
		default:
			usage();
			break;
		}
	argc -= optind;
	argv += optind;

	if (argc < 2)
		usage();

	fts_options = FTS_NOCHDIR | FTS_PHYSICAL;
	if (rflag) {
		if (Rflag)
			errx(1,
		    "the -R and -r options may not be specified together.");
		if (Hflag || Lflag || Pflag)
			errx(1,
	"the -H, -L, and -P options may not be specified with the -r option.");
		fts_options &= ~FTS_PHYSICAL;
		fts_options |= FTS_LOGICAL;
	}
	if (Rflag) {
		if (Hflag)
			fts_options |= FTS_COMFOLLOW;
		if (Lflag) {
			fts_options &= ~FTS_PHYSICAL;
			fts_options |= FTS_LOGICAL;
		}
	} else {
		fts_options &= ~FTS_PHYSICAL;
		fts_options |= FTS_LOGICAL;
	}

	myuid = getuid();

	/* Copy the umask for explicit mode setting. */
	myumask = umask(0);
	(void)umask(myumask);

	/* Save the target base in "to". */
	target = argv[--argc];
	if (strlen(target) >= sizeof(to.p_path))
		errx(1, "%s: name too long", target);
	(void)strcpy(to.p_path, target);
	to.p_end = to.p_path + strlen(to.p_path);
	if (to.p_path == to.p_end) {
		*to.p_end++ = '.';
		*to.p_end = '\0';
	}
	to.target_end = to.p_end;

	/* Set end of argument list for fts(3). */
	argv[argc] = NULL;

	/*
	 * Cp has two distinct cases:
	 *
	 * cp [-R] source target
	 * cp [-R] source1 ... sourceN directory
	 *
	 * In both cases, source can be either a file or a directory.
	 *
	 * In (1), the target becomes a copy of the source. That is, if the
	 * source is a file, the target will be a file, and likewise for
	 * directories.
	 *
	 * In (2), the real target is not directory, but "directory/source".
	 */
	r = stat(to.p_path, &to_stat);
	if (r == -1 && errno != ENOENT)
		err(1, "%s", to.p_path);
	if (r == -1 || !S_ISDIR(to_stat.st_mode)) {
		/*
		 * Case (1).  Target is not a directory.
		 */
		if (argc > 1) {
			usage();
			exit(1);
		}
		/*
		 * Need to detect the case:
		 *	cp -R dir foo
		 * Where dir is a directory and foo does not exist, where
		 * we want pathname concatenations turned on but not for
		 * the initial mkdir().
		 */
		if (r == -1) {
			if (rflag || (Rflag && (Lflag || Hflag)))
				stat(*argv, &tmp_stat);
			else
				lstat(*argv, &tmp_stat);

			if (S_ISDIR(tmp_stat.st_mode) && (Rflag || rflag))
				type = DIR_TO_DNE;
			else
				type = FILE_TO_FILE;
		} else
			type = FILE_TO_FILE;
	} else
		/*
		 * Case (2).  Target is a directory.
		 */
		type = FILE_TO_DIR;

	exit (copy(argv, type, fts_options));
}

char *
find_last_component(path)
	char *path;

{ 
	char *p;

	if ((p = strrchr(path, '/')) == NULL)
		p = path;
	else {
		/* Special case foo/ */
	        if (!*(p+1)) {
			while ((p >= path) &&
			       *p == '/')
				p--;

			while ((p >= path) &&
			       *p != '/')
				p--;
		}

		p++;
	}

	return (p);
}

int
copy(argv, type, fts_options)
	char *argv[];
	enum op type;
	int fts_options;
{
	struct stat to_stat;
	FTS *ftsp;
	FTSENT *curr;
	int base, dne, rval;
	size_t nlen;
	char *p, *target_mid;
#ifdef lint
	base = 0;
#endif
	if ((ftsp = fts_open(argv, fts_options, mastercmp)) == NULL)
		err(1, NULL);
	for (rval = 0; (curr = fts_read(ftsp)) != NULL;) {
		switch (curr->fts_info) {
		case FTS_NS:
		case FTS_ERR:
			warnx("%s: %s",
			    curr->fts_path, strerror(curr->fts_errno));
			rval = 1;
			continue;
		case FTS_DC:			/* Warn, continue. */
			warnx("%s: directory causes a cycle", curr->fts_path);
			rval = 1;
			continue;
		case FTS_DP:			/* Ignore, continue. */
			continue;
		}

		/*
		 * If we are in case (2) or (3) above, we need to append the 
		 * source name to the target name.  
		 */
		if (type != FILE_TO_FILE) {
			/*
			 * Need to remember the roots of traversals to create
			 * correct pathnames.  If there's a directory being
			 * copied to a non-existent directory, e.g.
			 *	cp -R a/dir noexist
			 * the resulting path name should be noexist/foo, not
			 * noexist/dir/foo (where foo is a file in dir), which
			 * is the case where the target exists.
			 *
			 * Also, check for "..".  This is for correct path
			 * concatenation for paths ending in "..", e.g.
			 *	cp -R .. /tmp
			 * Paths ending in ".." are changed to ".".  This is
			 * tricky, but seems the easiest way to fix the problem.
			 *
			 * XXX
			 * Since the first level MUST be FTS_ROOTLEVEL, base
			 * is always initialized.
			 */
			if (curr->fts_level == FTS_ROOTLEVEL)
				if (type != DIR_TO_DNE) {
					p = find_last_component(curr->fts_path);
					base = p - curr->fts_path;
					
					if (!strcmp(&curr->fts_path[base],
					    ".."))
						base += 1;
				} else
					base = curr->fts_pathlen;

			p = &curr->fts_path[base];
			nlen = curr->fts_pathlen - base;
			target_mid = to.target_end;
			if (*p != '/' && target_mid[-1] != '/')
				*target_mid++ = '/';
			*target_mid = '\0';
			if (target_mid - to.p_path + nlen >= MAXPATHLEN) {
				warnx("%s%s: name too long (not copied)",
				    to.p_path, p);
				rval = 1;
				continue;
			}
			(void)strncat(target_mid, p, nlen);
			to.p_end = target_mid + nlen;
			*to.p_end = '\0';
		}

		/* Not an error but need to remember it happened */
		if (stat(to.p_path, &to_stat) == -1)
			dne = 1;
		else {
			if (to_stat.st_dev == curr->fts_statp->st_dev &&
			    to_stat.st_ino == curr->fts_statp->st_ino) {
				warnx("%s and %s are identical (not copied).",
				    to.p_path, curr->fts_path);
				rval = 1;
				if (S_ISDIR(curr->fts_statp->st_mode))
					(void)fts_set(ftsp, curr, FTS_SKIP);
				continue;
			}
			if (!S_ISDIR(curr->fts_statp->st_mode) &&
			    S_ISDIR(to_stat.st_mode)) {
		warnx("cannot overwrite directory %s with non-directory %s",
				    to.p_path, curr->fts_path);
				rval = 1;
				continue;
			}
			dne = 0;
		}

		switch (curr->fts_statp->st_mode & S_IFMT) {
#ifdef S_IFLNK
		case S_IFLNK:
			if (copy_link(curr, !dne))
				rval = 1;
			break;
#endif
		case S_IFDIR:
			if (!Rflag && !rflag) {
				warnx("%s is a directory (not copied).",
				    curr->fts_path);
				(void)fts_set(ftsp, curr, FTS_SKIP);
				rval = 1;
				break;
			}
			/*
			 * If the directory doesn't exist, create the new
			 * one with the from file mode plus owner RWX bits,
			 * modified by the umask.  Trade-off between being
			 * able to write the directory (if from directory is
			 * 555) and not causing a permissions race.  If the
			 * umask blocks owner writes, we fail..
			 */
			if (dne) {
				if (mkdir(to.p_path,
				    curr->fts_statp->st_mode | S_IRWXU) < 0)
					err(1, "%s", to.p_path);
			} else if (!S_ISDIR(to_stat.st_mode)) {
				errno = ENOTDIR;
				err(1, "%s", to.p_path);
			}
			/*
			 * If not -p and directory didn't exist, set it to be
			 * the same as the from directory, unmodified by the 
			 * umask; arguably wrong, but it's been that way 
			 * forever.
			 */
			if (pflag && setfile(curr->fts_statp, 0))
				rval = 1;
			else if (dne)
				(void)chmod(to.p_path,
				    curr->fts_statp->st_mode);
			break;
#ifdef S_IFBLK
		case S_IFBLK:
#endif
		case S_IFCHR:
			if (Rflag) {
				if (copy_special(curr->fts_statp, !dne))
					rval = 1;
			} else
				if (copy_file(curr, dne))
					rval = 1;
			break;
		case S_IFIFO:
			if (Rflag) {
				if (copy_fifo(curr->fts_statp, !dne))
					rval = 1;
			} else
				if (copy_file(curr, dne))
					rval = 1;
			break;
		default:
			if (copy_file(curr, dne))
				rval = 1;
			break;
		}
	}
	if (errno)
		err(1, "fts_read");
	return (rval);
}

/*
 * mastercmp --
 *	The comparison function for the copy order.  The order is to copy
 *	non-directory files before directory files.  The reason for this
 *	is because files tend to be in the same cylinder group as their
 *	parent directory, whereas directories tend not to be.  Copying the
 *	files first reduces seeking.
 */
int
mastercmp(a, b)
	const FTSENT **a, **b;
{
	int a_info, b_info;

	a_info = (*a)->fts_info;
	if (a_info == FTS_ERR || a_info == FTS_NS || a_info == FTS_DNR)
		return (0);
	b_info = (*b)->fts_info;
	if (b_info == FTS_ERR || b_info == FTS_NS || b_info == FTS_DNR)
		return (0);
	if (a_info == FTS_D)
		return (-1);
	if (b_info == FTS_D)
		return (1);
	return (0);
}
